doc/SONiC_YANG_Model_Guidelines.md

```diff
@@ -120,39 +304,248 @@ container ACL_RULE {
 }
 ```
 
-### 4.  If an object is part of primary-key in ABNF.json, then it should be a key in YANG model. In YANG models, a primary-key from ABNF.json can be represented either as name of a Container object or as a key field in List object. Exception must be recorded in YANG Model with a comment in object field. To see an example of a comment, please refer Guideline # 7.
+### 12.  If an object is part of primary-key in ABNF.json, then it should be a key in YANG model. In YANG models, a primary-key from ABNF.json can be represented either as name of a Container object or as a key field in List object. Exception must be recorded in YANG Model with a comment in object field. To see an example of a comment, please refer Guideline #20. Though, key names are not stored in Redis DB, use the same key name as defined in ABNF schema.
 
-Example: VLAN_MEMBER dictionary in ABNF.json has both vlan-id and Port part of the key. So YANG model should have the same keys.
+Example: VLAN_MEMBER dictionary in ABNF.json has both vlan-id and ifname part of the key. So YANG model should have the same keys.
 
 #### ABNF
 ```
  "VLAN_MEMBER": {
-        "Vlan100|Ethernet0": {<<<< KEYS
+        "Vlan100|Ethernet0": { //<<<< KEYS
             "tagging_mode": "untagged"
         }
    }
 ```
+;Defines interfaces which are members of a vlan
+key                 = VLAN_MEMBER_TABLE:"Vlan"vlanid:ifname ;
+
 #### YANG
 ```
 container VLAN_MEMBER {
     description "VLAN_MEMBER part of config_db.json";
         list ..... {
-            key "vlanid port";<<<< KEYS
+            key "vlanid ifname";//<<<< KEYS
        }
 }
 ```
 
-### 5.  It is best to categorize YANG Modules based on a networking components. For Example, it is good to have separate modules for VLAN,  ACL, PORT and IP-ADDRESSES etc.
+### 13. If any key used in current table refers to other table, use leafref type for the key leaf definition. 
+
+Example : 
+#### ABNF
+
 ```
-sonic-acl.yang
-sonic-interface.yang
-sonic-port.yang
-sonic-vlan.yang
+key: ACL_RULE_TABLE:table_name:rule_name
+.....
+.....
+```
+#### YANG
+```
+.....
+container ACL_TABLE {
+	list ACL_TABLE_LIST {
+		key aclname;
+			
+		leaf aclname {
+			type string;
+		}
+		.....
+		.....
+	}
+}
+container ACL_RULE {
+	list ACL_RULE_LIST {
+		key "aclname rulename";
+		
+		leaf aclname {
+			type leafref {
+				path "../../../ACL_TABLE/ACL_TABLE_LIST/aclname"; //Refers to other table 'ACL_TABLE'
+			}
+		}
+		
+		leaf rulename {
+			type string;
+		}
+		.....
+		.....
+	}
+}
+```
+
+### 14. Typically the default ABNF key pattern is '{table_name}|{key1}|{key2}'. However, if needed use ``'*'`` for repetitive key pattern e.g. 'sonic-ext:key-pattern QUEUE|({ifname},)*|{qindex}'. 
+
```

Comment:
Not sure, if we need this. This could complicate the FW. I thought that we decided not to support repetitive keys at schema. If really needed, we could do some combining at the point of writing into DB.

Said that, can you please provide the ABNF schema entries that use repetitive keys. It would be few and we could vet rightaway, if multiple redis entries can be created instead of having one repeating keys,

e.g.:
  "PORT_QOS_MAP": {
        "Ethernet4,Ethernet8,Ethernet12,Ethernet16,Ethernet20,Ethernet24,Ethernet28,Ethernet32,Ethernet36,Ethernet40,Ethernet44,Ethernet48,Ethernet52,Ethernet56,Ethernet60,Ethernet64,Eth
ernet68,Ethernet72,Ethernet76,Ethernet80,Ethernet84,Ethernet88,Ethernet92,Ethernet96,Ethernet112,Ethernet116,Ethernet120,Ethernet124": {
            "tc_to_pg_map": "[TC_TO_PRIORITY_GROUP_MAP|AZURE]",
            "tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE]",
            "pfc_enable": "3,4",
            "pfc_to_queue_map": "[MAP_PFC_PRIORITY_TO_QUEUE|AZURE]",
            "dscp_to_tc_map": "[DSCP_TO_TC_MAP|AZURE]"
        }
    },

This could be split as 
  "PORT_QOS_MAP": {
        "Ethernet4": {
            "tc_to_pg_map": "[TC_TO_PRIORITY_GROUP_MAP|AZURE]",
            "tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE]",
            "pfc_enable": "3,4",
            "pfc_to_queue_map": "[MAP_PFC_PRIORITY_TO_QUEUE|AZURE]",
            "dscp_to_tc_map": "[DSCP_TO_TC_MAP|AZURE]"
        }
    },
  "PORT_QOS_MAP": {
        "Ethernet8": {
            "tc_to_pg_map": "[TC_TO_PRIORITY_GROUP_MAP|AZURE]",
            "tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE]",
            "pfc_enable": "3,4",
            "pfc_to_queue_map": "[MAP_PFC_PRIORITY_TO_QUEUE|AZURE]",
            "dscp_to_tc_map": "[DSCP_TO_TC_MAP|AZURE]"
        }
    },
  "PORT_QOS_MAP": {
        "Ethernet12": {
            "tc_to_pg_map": "[TC_TO_PRIORITY_GROUP_MAP|AZURE]",
            "tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE]",
            "pfc_enable": "3,4",
            "pfc_to_queue_map": "[MAP_PFC_PRIORITY_TO_QUEUE|AZURE]",
            "dscp_to_tc_map": "[DSCP_TO_TC_MAP|AZURE]"
        }
    },
...

Reply:
As per my finding, ABNF schema BUFFER_PG, PORT_QOS_MAP, QUEUE use repetitive keys.

Reply:
You are correct. As for BUFFER_PG, PORT_QO & QUEUE, I have checked with an SME (Wenda Ni) and he prefers the split.

Let us go by this simple way.

Reply:
Ok, I will remove it from the guideline.